### PR TITLE
triggers: refactor: add trap_common_t

### PR DESCRIPTION
**Pull request by YenHaoChen**
This PR depends on https://github.com/riscv-software-src/riscv-isa-sim/pull/1177.
I will rebase this PR after merging https://github.com/riscv-software-src/riscv-isa-sim/pull/1177

This PR adds trap_common_t to clean up itrigger and etrigger.

### CODE REVIEW

File: riscv/triggers.cc [hunk 1]
```diff
@@ -190,7 +190,8 @@ std::optional<match_result_t> mcontrol_common_t::detect_memory_access_match(proc
   if ((operation == triggers::OPERATION_EXECUTE && !execute) ||
       (operation == triggers::OPERATION_STORE && !store) ||
       (operation == triggers::OPERATION_LOAD && !load) ||
-      !mode_match(proc->get_state())) {
+      !mode_match(proc->get_state()) ||
+      !textra_match(proc)) {
```
Comment (scottj97) [inline riscv/triggers.cc]: Why not combine these two methods? They are always called together.

**Review by scottj97 — CHANGES_REQUESTED** [review]
Please break up 21b8a8f3286cc213a22b6eaaee913fa0d98fcb59 into smaller commits. First create an empty parent `trap_common_t` and then one-by-one move code/vars into the parent.

The commit message for b7a5882da2b533388128dc79d025b5d7012ae53f has an incorrect method name.

File: riscv/triggers.cc (hunk 1 above)
Comment (YenHaoChen) [reply riscv/triggers.cc]: I prefer to separate the ```mode_match()``` and ```textra_match()``` for the following reasons.
1. The ```(!mode_match(proc->get_state()) || !textra_match(proc))``` clarifies two independent checkings and, thus, has a higher readability than ```(!mode_and_textra_match(proc))```.
2. The current separated version of ```mode_match()``` is more elegant than ```mode_and_textra_match()```.
> bool trigger_t::mode_match(state_t * const state) const noexcept
> {
>   switch (state->prv) {
>     case PRV_M: return m;
>     case PRV_S: return state->v ? vs : s;
>     case PRV_U: return state->v ? vu : u;
>     default: assert(false);
>   }
> }

The ```mode_and_textra_match()``` will go with something like:

>   switch (state->prv) {
>     case PRV_M: if (!m) return false; break;
>     case PRV_S: if (!(state->v ? vs : s)) return false; break;
>     case PRV_U: if (!(state->v ? vu : u)) return false; break;
>     default: assert(false);
>   }
>   ...

3. The ```textra_match()``` is already a structurally complicated function. I want to keep it conceptually simple, i.e., checking textra CSR.

With the above concerns, how about adding a new function ```trigger_t::common_match()```, which performs common checks across all triggers? (prototyped in https://github.com/riscv-software-src/riscv-isa-sim/pull/1207/commits/590a36e487dc9bbb8d2c13bc4fd7e9af40cff9ad)
```
bool trigger_t::common_match(processor_t * const proc) const noexcept 
{ return mode_match(proc->get_state()) && textra_match(proc); }
```

Comment (scottj97) [reply riscv/triggers.cc]: I'm impressed with your reasoning. Let's leave it like it is.

**Review by scottj97 — APPROVED** [review]

**Pull request merged**